File: internal/db/cluster.go
package db

import (
	"context"
	"database/sql"
	"fmt"
)

func (db *DB) ComputeFeedSimilarity(ctx context.Context) error {
	_, err := db.ExecContext(ctx, `DELETE FROM feed_similarity`)
	if err != nil {
		return err
	}

	rows, err := db.QueryContext(ctx, `
		SELECT s1.feed_url, s2.feed_url, COUNT(*) AS overlap
		FROM subscriptions s1
		JOIN subscriptions s2 ON s1.user_did = s2.user_did AND s1.feed_url < s2.feed_url
		GROUP BY s1.feed_url, s2.feed_url
		HAVING overlap > 0
	`)
	if err != nil {
		return err
	}
	defer rows.Close()

	type pair struct {
		feedA   string
		feedB   string
		overlap int
	}
	var pairs []pair
	for rows.Next() {
		var p pair
		if err := rows.Scan(&p.feedA, &p.feedB, &p.overlap); err != nil {
			return err
		}
		pairs = append(pairs, p)
	}
	if err := rows.Err(); err != nil {
		return err
	}

	subCounts := make(map[string]int)
	for _, p := range pairs {
		subCounts[p.feedA] = 0
		subCounts[p.feedB] = 0
	}

	if len(subCounts) > 0 {
		countRows, err := db.QueryContext(ctx, `
			SELECT feed_url, COUNT(*) FROM subscriptions GROUP BY feed_url
		`)
		if err != nil {
			return err
		}
		for countRows.Next() {
			var feedURL string
			var count int
			if err := countRows.Scan(&feedURL, &count); err != nil {
				countRows.Close()
				return err
			}
			subCounts[feedURL] = count
		}
		countRows.Close()
	}

	for _, p := range pairs {
		total := subCounts[p.feedA] + subCounts[p.feedB] - p.overlap
		if total == 0 {
			continue
		}
		jaccard := float64(p.overlap) / float64(total)
		_, err := db.ExecContext(ctx, `
			INSERT INTO feed_similarity (feed_a, feed_b, jaccard, computed_at)
			VALUES (?, ?, ?, CURRENT_TIMESTAMP)
		`, p.feedA, p.feedB, jaccard)
		if err != nil {
			return err
		}
	}

	return nil
}

func (db *DB) ComputeUserSimilarity(ctx context.Context) error {
	_, err := db.ExecContext(ctx, `DELETE FROM user_similarity`)
	if err != nil {
		return err
	}

	rows, err := db.QueryContext(ctx, `
		SELECT s1.user_did, s2.user_did, COUNT(*) AS common
		FROM subscriptions s1
		JOIN subscriptions s2 ON s1.user_did < s2.user_did AND s1.feed_url = s2.feed_url
		GROUP BY s1.user_did, s2.user_did
		HAVING common > 0
	`)
	if err != nil {
		return err
	}
	defer rows.Close()

	type pair struct {
		userA  string
		userB  string
		common int
	}
	var pairs []pair
	for rows.Next() {
		var p pair
		if err := rows.Scan(&p.userA, &p.userB, &p.common); err != nil {
			return err
		}
		pairs = append(pairs, p)
	}
	if err := rows.Err(); err != nil {
		return err
	}

	subCounts := make(map[string]int)
	for _, p := range pairs {
		subCounts[p.userA] = 0
		subCounts[p.userB] = 0
	}

	if len(subCounts) > 0 {
		countRows, err := db.QueryContext(ctx, `
			SELECT user_did, COUNT(*) FROM subscriptions GROUP BY user_did
		`)
		if err != nil {
			return err
		}
		for countRows.Next() {
			var userDID string
			var count int
			if err := countRows.Scan(&userDID, &count); err != nil {
				countRows.Close()
				return err
			}
			subCounts[userDID] = count
		}
		countRows.Close()
	}

	for _, p := range pairs {
		total := subCounts[p.userA] + subCounts[p.userB] - p.common
		if total == 0 {
			continue
		}
		jaccard := float64(p.common) / float64(total)
		_, err := db.ExecContext(ctx, `
			INSERT INTO user_similarity (user_a, user_b, jaccard, common_feeds, computed_at)
			VALUES (?, ?, ?, ?, CURRENT_TIMESTAMP)
		`, p.userA, p.userB, jaccard, p.common)
		if err != nil {
			return err
		}
	}

	return nil
}

func (db *DB) ComputeFeedRecommendations(ctx context.Context, userDID string) error {
	_, err := db.ExecContext(ctx, `
		DELETE FROM user_feed_recommendations WHERE user_did = ?
	`, userDID)
	if err != nil {
		return err
	}

	rows, err := db.QueryContext(ctx, `
		SELECT
			CASE WHEN fs.feed_a IN (SELECT feed_url FROM subscriptions WHERE user_did = ?) THEN fs.feed_b ELSE fs.feed_a END AS recommended_feed,
			SUM(fs.jaccard) AS score
		FROM feed_similarity fs
		WHERE fs.feed_a IN (SELECT feed_url FROM subscriptions WHERE user_did = ?)
			OR fs.feed_b IN (SELECT feed_url FROM subscriptions WHERE user_did = ?)
		GROUP BY recommended_feed
		HAVING recommended_feed NOT IN (SELECT feed_url FROM subscriptions WHERE user_did = ?)
		ORDER BY score DESC
	`, userDID, userDID, userDID, userDID)
	if err != nil {
		return err
	}
	defer rows.Close()

	for rows.Next() {
		var feedURL string
		var score float64
		if err := rows.Scan(&feedURL, &score); err != nil {
			return err
		}
		_, err := db.ExecContext(ctx, `
			INSERT INTO user_feed_recommendations (user_did, feed_url, score, computed_at)
			VALUES (?, ?, ?, CURRENT_TIMESTAMP)
		`, userDID, feedURL, score)
		if err != nil {
			return err
		}
	}
	return rows.Err()
}

func (db *DB) GetFeedRecommendations(ctx context.Context, userDID string, limit int) ([]map[string]any, error) {
	rows, err := db.QueryContext(ctx, `
		SELECT r.feed_url, r.score, f.title, f.site_url, f.description, f.subscriber_count, f.favicon_url
		FROM user_feed_recommendations r
		JOIN feeds f ON f.feed_url = r.feed_url
		WHERE r.user_did = ?
		ORDER BY r.score DESC
		LIMIT ?
	`, userDID, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var results []map[string]any
	for rows.Next() {
		var feedURL string
		var score float64
		var title, siteURL, description sql.NullString
		var faviconURL sql.NullString
		var subCount int
		if err := rows.Scan(&feedURL, &score, &title, &siteURL, &description, &subCount, &faviconURL); err != nil {
			return nil, err
		}
		results = append(results, map[string]any{
			"feed_url":         feedURL,
			"score":            score,
			"title":            title,
			"site_url":         siteURL,
			"description":      description,
			"subscriber_count": subCount,
			"favicon_url":      faviconURL,
		})
	}
	return results, rows.Err()
}

func (db *DB) GetPeopleRecommendations(ctx context.Context, userDID string, limit int) ([]map[string]any, error) {
	rows, err := db.QueryContext(ctx, fmt.Sprintf(`
		SELECT
			CASE WHEN us.user_a = ? THEN us.user_b ELSE us.user_a END AS recommended_user,
			us.jaccard, us.common_feeds,
			u.handle, u.display_name, u.avatar_url
		FROM user_similarity us
		JOIN users u ON u.did = CASE WHEN us.user_a = ? THEN us.user_b ELSE us.user_a END
		WHERE (us.user_a = ? OR us.user_b = ?)
		  AND u.handle IS NOT NULL AND u.handle != ''
		ORDER BY us.jaccard DESC
		LIMIT %d
	`, limit), userDID, userDID, userDID, userDID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var results []map[string]any
	for rows.Next() {
		var recUser, handle string
		var jaccard float64
		var commonFeeds int
		var displayName, avatarURL sql.NullString
		if err := rows.Scan(&recUser, &jaccard, &commonFeeds, &handle, &displayName, &avatarURL); err != nil {
			return nil, err
		}
		results = append(results, map[string]any{
			"did":          recUser,
			"jaccard":      jaccard,
			"common_feeds": commonFeeds,
			"handle":       handle,
			"display_name": displayName.String,
			"avatar_url":   avatarURL.String,
		})
	}
	return results, rows.Err()
}

func (db *DB) GetSimilarFeeds(ctx context.Context, feedURL string, limit int) ([]*Feed, error) {
	rows, err := db.QueryContext(ctx, `
		SELECT f.feed_url, f.title, f.site_url, f.description, f.feed_type,
			f.last_fetched_at, f.last_error, f.subscriber_count, f.etag, f.last_modified,
			f.fetch_interval_minutes, f.next_fetch_at, f.consecutive_empty_fetches, f.error_count, f.favicon_url
		FROM feed_similarity fs
		JOIN feeds f ON f.feed_url = CASE WHEN fs.feed_a = ? THEN fs.feed_b ELSE fs.feed_a END
		WHERE fs.feed_a = ? OR fs.feed_b = ?
		ORDER BY fs.jaccard DESC
		LIMIT ?
	`, feedURL, feedURL, feedURL, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var feeds []*Feed
	for rows.Next() {
		f := &Feed{}
		if err := rows.Scan(&f.FeedURL, &f.Title, &f.SiteURL, &f.Description, &f.FeedType,
			&f.LastFetchedAt, &f.LastError, &f.SubscriberCount, &f.Etag, &f.LastModified,
			&f.FetchIntervalMinutes, &f.NextFetchAt, &f.ConsecutiveEmptyFetches, &f.ErrorCount, &f.FaviconURL); err != nil {
			return nil, err
		}
		feeds = append(feeds, f)
	}
	return feeds, rows.Err()
}

type ArticleRecommendation struct {
	ArticleID int64
	Title     string
	URL       string
	FeedURL   string
	FeedTitle string
	Author    string
	Summary   string
	Published sql.NullTime
	Score     float64
}

func (db *DB) GetArticleRecommendations(ctx context.Context, userDID string, limit int) ([]*ArticleRecommendation, error) {
	rows, err := db.QueryContext(ctx, `
		SELECT a.id, a.title, COALESCE(a.url, ''), r.feed_url, COALESCE(f.title, ''),
			COALESCE(a.author, ''), COALESCE(a.summary, ''), a.published, r.score
		FROM user_article_recommendations r
		JOIN articles a ON a.feed_url = r.feed_url AND a.url = r.article_url
		LEFT JOIN feeds f ON f.feed_url = r.feed_url
		WHERE r.user_did = ?
		ORDER BY r.score DESC
		LIMIT ?
	`, userDID, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var recs []*ArticleRecommendation
	for rows.Next() {
		rec := &ArticleRecommendation{}
		if err := rows.Scan(&rec.ArticleID, &rec.Title, &rec.URL, &rec.FeedURL, &rec.FeedTitle,
			&rec.Author, &rec.Summary, &rec.Published, &rec.Score); err != nil {
			return nil, err
		}
		recs = append(recs, rec)
	}
	return recs, rows.Err()
}

File: spike/nim/atproto/cluster.nim
## The similarity computations behind recommendations, ported from
## internal/cluster/jaccard.go and the pure helpers of scoring.go.
##
## Two kinds of similarity are precomputed on a schedule, because both are
## O(pairs) and neither needs to be fresh to the second:
##
##   feed_similarity   which feeds are subscribed to by the same people
##   user_similarity   which people subscribe to the same feeds
##
## Both are time-decayed. An overlap from three years ago says much less
## about what someone wants now than one from last week, and without decay a
## long-lived account's early subscriptions dominate its recommendations
## forever.
##
## The decay constant is `exp(-0.023 * days)`, a half-life of about a month
## (ln 2 / 0.023 ≈ 30). It is written inline in the SQL rather than named,
## the same way the Go implementation writes it.

import std/[algorithm, math, options, random]
import ./gleandb
import ./sqlite

type
  Config* = object
    followBoost*: float
    likesWeight*: float
    tagsWeight*: float
    descriptionWeight*: float

  SignalWeights* = object
    wSub*: float
    wLike*: float
    wTag*: float
    wSocial*: float
    wPop*: float
    wCategory*: float
    wContent*: float

  FeedRecommendation* = object
    feedUrl*: string
    title*: string
    siteUrl*: string
    description*: string
    subscriberCount*: int
    faviconUrl*: string
    score*: float

  PersonRecommendation* = object
    did*: string
    handle*: string
    displayName*: string
    avatarUrl*: string
    commonFeeds*: int
    commonLikes*: int
    commonTags*: int
    isFollowed*: bool
    score*: float

const
  ## Half-life of roughly 30 days: ln(2) / 0.023.
  DecayPerDay* = 0.023

proc defaultConfig*(): Config =
  Config(followBoost: 0.5, likesWeight: 0.3, tagsWeight: 0.2,
         descriptionWeight: 0.15)

proc defaultWeights*(): SignalWeights =
  SignalWeights(wSub: 1.0, wLike: 0.5, wTag: 0.3, wSocial: 0.7,
                wPop: 0.2, wCategory: 0.4, wContent: 0.4)

proc decay*(ageDays: float): float =
  ## The weight an event of this age still carries.
  ##
  ## Exposed so the decay can be checked directly; the queries compute it in
  ## SQL because pulling every like into the application to weight it would
  ## defeat the point of precomputing.
  exp(-DecayPerDay * ageDays)

# --- similarity ------------------------------------------------------------

## Both computations stage into a temp table and swap in a second
## transaction. The first pass is the expensive one and holds no lock on the
## live table; readers keep seeing the previous generation until the swap,
## rather than an empty table for the length of the computation.

proc computeFeedSimilarity*(g: GleanDb) =
  ## For each pair of feeds, how strongly their subscriber sets overlap.
  ##
  ## The numerator is a decayed count of shared subscribers and the
  ## denominator is the union of the two subscriber counts, which makes this
  ## a Jaccard index with recency folded into the intersection.
  g.db.transaction:
    g.db.exec """CREATE TEMP TABLE IF NOT EXISTS _feed_sim_staging (
      feed_a TEXT NOT NULL, feed_b TEXT NOT NULL, jaccard REAL NOT NULL,
      PRIMARY KEY (feed_a, feed_b))"""
    g.db.exec "DELETE FROM _feed_sim_staging"
    g.db.exec """
      INSERT INTO _feed_sim_staging (feed_a, feed_b, jaccard)
      SELECT feed_a, feed_b, jaccard FROM (
        SELECT
          s1.feed_url AS feed_a,
          s2.feed_url AS feed_b,
          SUM(EXP(-0.023 * CAST(julianday('now') -
                                julianday(MIN(s1.added_at, s2.added_at)) AS REAL)))
          / NULLIF(f1.subscriber_count + f2.subscriber_count
                   - CAST(COUNT(*) AS REAL), 0) AS jaccard
        FROM articles.subscriptions s1
        -- feed_url < feed_url keeps one row per unordered pair, and drops
        -- the self-pair without a separate predicate.
        JOIN articles.subscriptions s2
          ON s1.user_did = s2.user_did AND s1.feed_url < s2.feed_url
        JOIN articles.feeds f1 ON f1.feed_url = s1.feed_url
        JOIN articles.feeds f2 ON f2.feed_url = s2.feed_url
        WHERE s1.added_at IS NOT NULL AND s2.added_at IS NOT NULL
        GROUP BY s1.feed_url, s2.feed_url
      ) WHERE jaccard IS NOT NULL"""

  g.db.transaction:
    g.db.exec "DELETE FROM recs.feed_similarity"
    g.db.exec """INSERT INTO recs.feed_similarity (feed_a, feed_b, jaccard)
                 SELECT feed_a, feed_b, jaccard FROM _feed_sim_staging"""

proc computeUserSimilarity*(g: GleanDb) =
  ## For each pair of people, how strongly their subscriptions overlap.
  ##
  ## Undecayed, unlike feed similarity: a subscription is a standing choice
  ## rather than an event, so its age says little. Shared *likes* are decayed
  ## separately below, because those are events.
  g.db.transaction:
    g.db.exec """CREATE TEMP TABLE IF NOT EXISTS _user_sim_staging (
      user_a TEXT NOT NULL, user_b TEXT NOT NULL, jaccard REAL NOT NULL,
      common_feeds INT NOT NULL DEFAULT 0,
      common_likes INT NOT NULL DEFAULT 0,
      common_tags INT NOT NULL DEFAULT 0,
      PRIMARY KEY (user_a, user_b))"""
    g.db.exec "DELETE FROM _user_sim_staging"
    g.db.exec """
      INSERT INTO _user_sim_staging
        (user_a, user_b, jaccard, common_feeds, common_likes, common_tags)
      SELECT
        s1.user_did, s2.user_did,
        CAST(COUNT(*) AS REAL) / (
          (SELECT COUNT(*) FROM articles.subscriptions WHERE user_did = s1.user_did) +
          (SELECT COUNT(*) FROM articles.subscriptions WHERE user_did = s2.user_did) -
          CAST(COUNT(*) AS REAL)
        ),
        COUNT(*), 0, 0
      FROM articles.subscriptions s1
      JOIN articles.subscriptions s2
        ON s1.feed_url = s2.feed_url AND s1.user_did < s2.user_did
      GROUP BY s1.user_did, s2.user_did"""

    # Shared likes, decayed on both sides: two people who liked the same
    # thing years apart have less in common than two who liked it the same
    # week, and the product of the two decays expresses that.
    #
    # ROUND, not CAST AS INTEGER. The Go implementation truncates, and since
    # each decay factor is at most 1 their product is always just under it --
    # so a pair sharing one recent like sums to 0.9999 and stores 0. That
    # number is shown to readers as the reason for a recommendation ("N
    # shared likes"), so truncation makes a real overlap read as none.
    g.db.exec """CREATE TEMP TABLE IF NOT EXISTS _likes_overlap (
      user_a TEXT, user_b TEXT, common INT, PRIMARY KEY(user_a, user_b))"""
    g.db.exec "DELETE FROM _likes_overlap"
    g.db.exec """
      INSERT INTO _likes_overlap (user_a, user_b, common)
      SELECT l1.author_did, l2.author_did,
        CAST(ROUND(SUM(
            EXP(-0.023 * CAST(julianday('now') - julianday(l1.created_at) AS REAL))
          * EXP(-0.023 * CAST(julianday('now') - julianday(l2.created_at) AS REAL))
        )) AS INTEGER)
      FROM articles.likes l1
      JOIN articles.likes l2
        ON l1.feed_url = l2.feed_url AND l1.article_url = l2.article_url
       AND l1.author_did < l2.author_did
      WHERE l1.created_at IS NOT NULL AND l2.created_at IS NOT NULL
      GROUP BY l1.author_did, l2.author_did"""

    # A pair that shares likes but no feeds has no staging row yet, so this
    # inserts before it updates.
    g.db.exec """
      INSERT INTO _user_sim_staging (user_a, user_b, jaccard, common_feeds,
                                     common_likes, common_tags)
      SELECT o.user_a, o.user_b, 0, 0, o.common, 0
      FROM _likes_overlap o
      WHERE NOT EXISTS (SELECT 1 FROM _user_sim_staging s
                        WHERE s.user_a = o.user_a AND s.user_b = o.user_b)"""
    g.db.exec """
      UPDATE _user_sim_staging SET common_likes = (
        SELECT o.common FROM _likes_overlap o
        WHERE o.user_a = _user_sim_staging.user_a
          AND o.user_b = _user_sim_staging.user_b)
      WHERE EXISTS (SELECT 1 FROM _likes_overlap o
                    WHERE o.user_a = _user_sim_staging.user_a
                      AND o.user_b = _user_sim_staging.user_b)"""

  g.db.transaction:
    g.db.exec "DELETE FROM recs.user_similarity"
    g.db.exec """INSERT INTO recs.user_similarity
                   (user_a, user_b, jaccard, common_feeds, common_likes, common_tags)
                 SELECT user_a, user_b, jaccard, common_feeds, common_likes,
                        common_tags
                 FROM _user_sim_staging"""

proc feedSimilarity*(g: GleanDb, feedA, feedB: string): float =
  ## Stored pairs are ordered, so look up whichever way round the caller asks.
  let (a, b) = if feedA < feedB: (feedA, feedB) else: (feedB, feedA)
  let row = g.db.queryFirst(
    "SELECT jaccard FROM recs.feed_similarity WHERE feed_a = ? AND feed_b = ?",
    [p(a), p(b)], proc(r: Row): float = r.f(0))
  if row.isSome: row.get else: 0.0

proc userSimilarity*(g: GleanDb, userA, userB: string): float =
  let (a, b) = if userA < userB: (userA, userB) else: (userB, userA)
  let row = g.db.queryFirst(
    "SELECT jaccard FROM recs.user_similarity WHERE user_a = ? AND user_b = ?",
    [p(a), p(b)], proc(r: Row): float = r.f(0))
  if row.isSome: row.get else: 0.0

# --- scoring helpers -------------------------------------------------------

proc normalizeScores*(scores: var openArray[float]) =
  ## Rescale to 0..1.
  ##
  ## Scores from different signals are not comparable in magnitude -- a
  ## like-overlap sum and a log-popularity term live on unrelated scales --
  ## so they are normalised before being blended or ranked against each
  ## other. A single item, or a set where every score is equal, is left
  ## alone: there is no spread to stretch, and dividing by zero would turn
  ## the whole list into NaN.
  if scores.len < 2: return
  var lo = scores[0]
  var hi = scores[0]
  for s in scores:
    lo = min(lo, s)
    hi = max(hi, s)
  if hi == lo: return
  let span = hi - lo
  for i in 0 ..< scores.len:
    scores[i] = (scores[i] - lo) / span

proc normalizeFeedScores*(recs: var seq[FeedRecommendation]) =
  var s = newSeq[float](recs.len)
  for i, r in recs: s[i] = r.score
  normalizeScores(s)
  for i in 0 ..< recs.len: recs[i].score = s[i]

proc normalizePersonScores*(recs: var seq[PersonRecommendation]) =
  var s = newSeq[float](recs.len)
  for i, r in recs: s[i] = r.score
  normalizeScores(s)
  for i in 0 ..< recs.len: recs[i].score = s[i]

proc samplePeople*(pool: seq[PersonRecommendation], half: int,
                   rng: var Rand): seq[PersonRecommendation] =
  ## Up to `half` from inside the reader's network and `half` from outside.
  ##
  ## The split is the point: ranking people purely by similarity returns the
  ## reader's existing network, which they already have. Reserving half the
  ## slots for strangers is what makes this discovery rather than a mirror.
  ##
  ## Shuffled rather than taken in order so the same faces do not appear
  ## every time -- the pool changes far more slowly than the reader revisits.
  var inNet, outNet: seq[PersonRecommendation]
  for p in pool:
    if p.isFollowed: inNet.add p else: outNet.add p

  rng.shuffle(inNet)
  rng.shuffle(outNet)
  if inNet.len > half: inNet.setLen(half)
  if outNet.len > half: outNet.setLen(half)
  inNet & outNet

proc rankByScore*[T](items: var seq[T]) =
  ## Highest first, which is what every recommendation listing wants.
  items.sort(proc(a, b: T): int = cmp(b.score, a.score))
